compiler/rustc_mir_transform/src/value_numbering.rs

```diff
@@ -0,0 +1,907 @@
+            return false;
+        }
+
+        let mut ty_kind = local_decl.ty.kind();
+        for place_elem in &place.projection {
+            match *place_elem {
+                ProjectionElem::Deref => match ty_kind {
+                    ty::Ref(_, inner_ty, Mutability::Not) => {
+                        ty_kind = inner_ty.kind();
+                    }
+                    _ => return false,
+                },
+                ProjectionElem::Field(_, _) => {
+                    match ty_kind {
+                        ty::Adt(ref adt, _)
+                            if Some(adt.did) != self.tcx.lang_items().unsafe_cell_type() =>
+                        {
+                            // The field of an `UnsafeCell` is considered mutable.
+                        }
+                        _ => return false,
+                    }
+                }
+                // FIXME(pcwalton): Handle some of these.
+                ProjectionElem::ConstantIndex { .. }
+                | ProjectionElem::Downcast(_, _)
+                | ProjectionElem::Index(_)
+                | ProjectionElem::Subslice { .. } => return false,
+            }
+        }
+
+        true
+    }
+}
+
+// A very conservative alias analysis that detects when locals might have been mutated by a
+// statement.
+//
+// FIXME(pcwalton): Improve precision.
+struct VnAliasAnalysis<'a, 'vn, 'tcx> {
+    local_decls: &'a LocalDecls<'tcx>,
+    locals: &'vn mut FxHashMap<Local, VnLocalInfo>,
+}
+
+impl<'a, 'vn, 'tcx> VnAliasAnalysis<'a, 'vn, 'tcx> {
+    fn new(
+        local_decls: &'a LocalDecls<'tcx>,
+        locals: &'vn mut FxHashMap<Local, VnLocalInfo>,
+    ) -> Self {
+        Self { local_decls, locals }
+    }
+}
+
+impl<'a, 'vn, 'tcx> Visitor<'tcx> for VnAliasAnalysis<'a, 'vn, 'tcx> {
+    fn visit_place(
+        &mut self,
+        place: &Place<'tcx>,
+        place_context: PlaceContext,
+        location: Location,
+    ) {
+        match place_context {
+            PlaceContext::NonUse(_) | PlaceContext::NonMutatingUse(_) => return,
+            PlaceContext::MutatingUse(_) => {}
+        }
+
+        let mut local_info = self.locals.get_mut(&place.local).unwrap();
+        local_info.statement_index_of_last_kill = location.statement_index;
+        local_info.current_live_range_index += 1;
+
+        // Kill every mutable local if we see any projection (most importantly, a dereference).
+        // This includes mutating through references, etc.
+        //
+        // FIXME(pcwalton): This is super conservative.
+        if !place.projection.is_empty() {
+            for (local, local_info) in self.locals.iter_mut() {
+                if *local != place.local && self.local_decls[*local].mutability == Mutability::Mut {
+                    local_info.statement_index_of_last_kill = location.statement_index;
+                    local_info.current_live_range_index += 1;
+                }
+            }
+        }
```

Comment:
The idea here is that we wait until an indirect write to mark all locals that are (possibly) mutably borrowed as possibly overwritten, correct? Do we do the same for function calls? Those can also write to locals indirectly.

Reply:
Also, any reason you don't check for `!place.is_indirect()` instead of ensuring that `projection` is empty? This is more conservative but also less clear.

Reply:
Oh, one more thing. The precedent for indirect mutation is to traverse the MIR to looking for any locals that are mutably borrowed at any point. I'm not sure if you need to look at the mutability of a `LocalDecl` separately, but maybe doing it that way helps you avoid changing that failing test case?

Reply:
This is *local* value numbering, local to a basic block. Function calls and drops create new basic blocks, so we don't have to handle them separately.